Display conversion progress

**Required skills:** Python

**Difficulty:** Easy

For the initial asset conversion, openage provides textual output to signal which stage of conversion has been reached. The output usually looks like this:

```
INFO [py] converting metadata
INFO [py] [0] palette
INFO [py] [1] empires.dat
INFO [py] Starting conversion...
INFO [py] Extracting Genie data...
INFO [py] Creating API-like objects...
INFO [py] Linking API-like objects...
INFO [py] Generating auxiliary objects...
INFO [py] Creating nyan objects...
INFO [py] Creating requests for media export...
INFO [py] Starting export...
INFO [py] Dumping info file...
INFO [py] Dumping data files...
INFO [py] Exporting media files...
INFO [py] Dumping metadata files...
INFO [py] [2] player color palette
INFO [py] [3] terminal color palette
INFO [py] [4] game specification files
INFO [py] asset conversion complete; asset version: 7
```

However, some of these stages can take a while to complete. For example, the whole `Exporting media files...` stage can take up to 20 minutes. For these stages, it would be nice to have more dynamic updates in form of a progress bar. This would also give much more feedback to users on what's being done at the moment.

The stages that would benefit the most from a progress bar are:

- [ ] `empires.dat` file parsing ([See code](https://github.com/SFTtech/openage/blob/master/openage/convert/entity_object/conversion/genie_structure.py#L61))
- [ ] `Creating nyan objects...` ([See code](https://github.com/SFTtech/openage/blob/master/openage/convert/processor/conversion/aoc/nyan_subprocessor.py#L71))
- [ ] `Exporting media files...`  ([See code](https://github.com/SFTtech/openage/blob/master/openage/convert/processor/export/media_exporter.py#L22))

We provide a formatter for the progress bar in our tools: https://github.com/SFTtech/openage/blob/master/openage/util/strings.py#L103


**Further reading**:

* [Converter architecture](https://github.com/SFTtech/openage/blob/master/doc/code/converter/architecture_overview.md)
* [Converter workflow](https://github.com/SFTtech/openage/blob/master/doc/code/converter/workflow.md)

@heinezen have you seen this error before during configuration? I followed the instructions on https://github.com/SFTtech/openage/blob/master/doc/build_instructions/macos.md

```

CMake Warning at buildsystem/HandleCXXOptions.cmake:75 (message):
  GNU gold linker isn't available, using the default system linker.
Call Stack (most recent call first):
  buildsystem/HandleCXXOptions.cmake:107 (try_enable_gold_linker)
  CMakeLists.txt:172 (include)




The compiler doesn't support required C++20 features:
  * Concepts
  * Default comparisons
The following versions support these features:
  * clang++ >= 10
  * g++ >= 10
  * Microsoft Visual Studio 2019 >= 16.8
Please upgrade your compiler to build openage.


CMake Error at buildsystem/CheckCompilerFeatures.cmake:57 (message):
  aborting
Call Stack (most recent call first):
  CMakeLists.txt:173 (include)
```

--- a/copying.md
+++ b/copying.md
@@ -141,6 +141,7 @@ _the openage authors_ are:
 | Damien Lejay                | dlejay                      | lejay à paracompact dawt space                    |
 | Talha Aamir                 | sarcxd                      | sarcxd à gmail dawt com                           |
 | Matthias Geiger             | CountOmega                  | matthias dawt geiger1024 à outlook dawt com       |
+| Yuvraj Tetarwal             | YuviTz1                     | yuvi56789 à gmail dawt com                        |
 
 If you're a first-time committer, add yourself to the above list. This is not
 just for legal reasons, but also to keep an overview of all those nicknames.

--- a/openage/convert/processor/export/media_exporter.py
+++ b/openage/convert/processor/export/media_exporter.py
@@ -17,6 +17,7 @@ from openage.convert.service.export.load_media_cache import load_media_cache
 from openage.convert.value_object.read.media.blendomatic import Blendomatic
 from openage.convert.value_object.read.media_types import MediaType
 from openage.log import dbg, get_loglevel
+from openage.util.strings import format_progress
 
 if typing.TYPE_CHECKING:
     from argparse import Namespace
@@ -69,22 +70,30 @@ class MediaExporter:
                 kwargs["palettes"] = args.palettes
                 kwargs["compression_level"] = args.compression_level
                 export_func = MediaExporter._export_terrain
+                print("-- Exporting terrain files...")
 
             elif media_type is MediaType.GRAPHICS:
                 kwargs["palettes"] = args.palettes
                 kwargs["compression_level"] = args.compression_level
                 kwargs["cache_info"] = cache_info
                 export_func = MediaExporter._export_graphics
+                print("-- Exporting graphics files...")
 
             elif media_type is MediaType.SOUNDS:
                 export_func = MediaExporter._export_sound
+                print("-- Exporting sound files...")
 
             elif media_type is MediaType.BLEND:
                 kwargs["blend_mode_count"] = args.blend_mode_count
                 export_func = MediaExporter._export_blend
+                print("-- Exporting blend files...")
 
+            export_count = 1
+            total_count = len(cur_export_requests)
             for request in cur_export_requests:
                 export_func(request, sourcedir, exportdir, **kwargs)
+                print(f"-- Files done: {format_progress(export_count, total_count)}", end = "\r", flush = True)
+                export_count += 1
 
         if args.debug_info > 5:
             cachedata = {}

--- a/openage/convert/processor/export/media_exporter.py
+++ b/openage/convert/processor/export/media_exporter.py
@@ -92,7 +92,8 @@ class MediaExporter:
             total_count = len(cur_export_requests)
             for request in cur_export_requests:
                 export_func(request, sourcedir, exportdir, **kwargs)
-                print(f"-- Files done: {format_progress(export_count, total_count)}", end = "\r", flush = True)
+                print(f"-- Files done: {format_progress(export_count, total_count)}",
+                      end = "\r", flush = True)
                 export_count += 1
 
         if args.debug_info > 5:

--- a/openage/convert/processor/export/media_exporter.py
+++ b/openage/convert/processor/export/media_exporter.py
@@ -88,13 +88,11 @@ class MediaExporter:
                 export_func = MediaExporter._export_blend
                 print("-- Exporting blend files...")
 
-            export_count = 1
             total_count = len(cur_export_requests)
-            for request in cur_export_requests:
+            for count, request in enumerate(cur_export_requests, start = 1):
                 export_func(request, sourcedir, exportdir, **kwargs)
-                print(f"-- Files done: {format_progress(export_count, total_count)}",
+                print(f"-- Files done: {format_progress(count, total_count)}",
                       end = "\r", flush = True)
-                export_count += 1
 
         if args.debug_info > 5:
             cachedata = {}

